### PR TITLE
Gemma-scope-2 Neuronpedia aliases, in-place decoder-norm fold, and an activations-store device seam

### PR DESCRIPTION
**Pull request by speediedan**
> Thank you so much to the SAELens maintainers and community for making this foundationally valuable
> contribution to the open-source interpretability ecosystem! I'm using it extensively in the
> downstream world-model analysis framework I'm building (Interpretune) and couldn't appreciate your
> work more. The small seams proposed below come directly out of that heavy use.

## Overview

This PR carries the small SAELens surface that the coordinated **Scalable Dashboards** work depends
on: Neuronpedia source-set alias metadata for the gemma-scope-2 instruction-tuned transcoder
releases, a parameter-storage-preserving form of the decoder-norm fold, and an opt-in device seam on
`ActivationsStore.get_batch_tokens()`. SAELens plays a supporting role in the wave — the companion
PRs are SAEDashboard (columnar generation lane), neuronpedia (Python import utilities), and a
[coordination umbrella in interpretune](https://github.com/speediedan/interpretune/issues/231) linking the set (soft merge order: **SAELens → SAEDashboard →
neuronpedia**; nothing here hard-depends on the others).

> [!NOTE]
> **Ownership framing**: the upstream repos — principally Neuronpedia, together with SAEDashboard —
> own the core dashboard APIs and generation protocol and will continue to own them. Interpretune
> provides one example generation/import orchestration pipeline demonstrating the proposed
> scalability and example-aligned customizability. This PR does not redefine upstream interfaces or
> ownership.

### 1. `pretrained_saes.yaml`: gemma-scope-2 Neuronpedia aliases + registry repair

Adds `neuronpedia:` aliases for the gemma-scope-2 instruction-tuned transcoder releases:

- `gemma-scope-2-4b-it-transcoders-all` `width_16k_l0_small_affine` → `gemma-3-4b-it/<layer>-gemmascope-2-transcoder-16k`
- `gemma-scope-2-1b-it-transcoders-all` `width_16k_l0_small_affine` → `gemma-3-1b-it/<layer>-gemmascope-2-transcoder-16k`
- `gemma-scope-2-1b-it-transcoders-all` `width_262k_l0_small_affine` → `gemma-3-1b-it/<layer>-gemmascope-2-transcoder-262k`

and repairs pre-existing structural damage in the `gemma-scope-2-12b-pt-transcoders-all`
`width_262k_l0_big` block (missing `- id:` keys), keeping the registry valid.

**Why**: the SAE rows already exist upstream, but Neuronpedia inference resolves source sets through
the `neuronpedia_id` alias metadata — without these aliases a local/hosted inference server cannot
translate e.g. `SAE_SETS=["gemmascope-2-transcoder-262k"]` into SAE ids for these releases (a
runtime registry probe returns `no_match`). SAELens is the single source of truth for this metadata;
synthesizing aliases in Neuronpedia's config layer was tried and reverted.

### 2. In-place `fold_W_dec_norm` (parameter-storage preservation)

`fold_and_get_W_dec_norm` (the shared helper from the "DRY fold decoder norms" refactor, #705) now
applies the fold with in-place `div_`/`mul_` on the existing parameter tensors instead of rebinding
`.data`; the gated-SAE `b_gate`/`b_mag` handling does the same. Numerics are unchanged (same clamp,
same order of operations).

**Why**: rebinding `.data` swaps the backing storage, silently invalidating external references held
by optimizers, hooks, or downstream tools that captured views of the parameters before folding —
SAEDashboard's runner is one such caller.

### 3. Activations-store device-transfer override

`ActivationsStore.get_batch_tokens()` gains an optional `move_to_model_device` argument (default
`True`, existing behavior unchanged), letting callers keep the stacked token batch on the
activations-store device; the seam is preserved across store refills.

**Why**: SAEDashboard's `NeuronpediaRunner.generate_tokens()` deduplicates and shuffles prompt
batches before model execution and calls `get_batch_tokens(move_to_model_device=False)` so staging
happens once, on the right device — downstream callers cannot request this cleanly against the
released surface without the seam.

## Compatibility / Testing

- Commit 1: runtime Neuronpedia-directory probes return the full expected layer counts per release
  (e.g. 26 `google/gemma-3-1b-it` rows for each aliased source set).
- Commit 2: new `test_fold_W_dec_norm_keeps_parameter_storage_in_place`, parametrized over all
  foldable architectures, pins `data_ptr()` stability for `W_dec`, `W_enc`, and `b_enc` (when
  present) across the fold; full `test_standard_sae.py` / `test_gated_sae.py` / `test_sae.py`
  suites pass.
- Commit 3: activations-store unit tests extended; 190 tests across the touched suites pass.
- Rebase note: upstream's fold-helper refactor (#705) landed between our fork point and the rebase —
  the in-place change is expressed inside the upstream helper structure (not a parallel
  implementation) and the storage test sits alongside upstream's relocated fold tests.
- Downstream validation: these three seams have run under the Scalable Dashboards production +
  benchmark workloads (26-layer gemma-3-1b-it 262k-transcoder dashboard generation; three-way parity
  suite) for several weeks — [evidence package](https://github.com/speediedan/interpretune/blob/streamlined-streamable-dashboard-generation-phase-1/docs/benchmark_artifacts/full_20260727/benchmark_summary.md) linked from the [coordination umbrella](https://github.com/speediedan/interpretune/issues/231).

## Coordination

Part of the coordinated upstream **Scalable Dashboards** PR set (Wave 1): SAEDashboard (columnar
generation lane), neuronpedia (import utilities + ecosystem overview), circuit-tracer#79 (existing),
with the [coordination umbrella in interpretune](https://github.com/speediedan/interpretune/issues/231) hosting the reproduction quickstart, benchmark
evidence, and the Wave 1/Wave 2 roadmap. Soft merge order SAELens → SAEDashboard → neuronpedia
(SAELens first unblocks the neuronpedia inference pin flip to a released SAELens floor).
@chanind 

---

## Verifying this PR yourself

Runner: `make test` = `poetry run pytest -v --cov=sae_lens/ tests`.

| Claim | Verify it | Cost |
| --- | --- | --- |
| In-place fold preserves parameter storage | `poetry run pytest tests/saes/test_standard_sae.py::test_fold_W_dec_norm_keeps_parameter_storage_in_place` → 3 passed (parametrized over standard/gated/jumprelu; asserts `data_ptr()` stability) | seconds, CPU, offline |
| No fold regressions | `poetry run pytest tests/saes/test_standard_sae.py tests/saes/test_gated_sae.py tests/saes/test_jumprelu_sae.py -k fold` → 11 passed | seconds |
| `move_to_model_device` seam | `poetry run pytest tests/training/test_activations_store.py::test_get_batch_tokens_can_skip_model_device_transfer` — **CUDA-gated; silently skips on CPU-only machines** | seconds, GPU |
| Surrounding activations-store behavior | `poetry run pytest tests/training/test_activations_store.py -k "get_batch_tokens or input_token_device"` → 10 passed | seconds |
| Alias metadata resolves | the snippet below, offline (`HF_HUB_OFFLINE=1`), no auth or gated weights needed | seconds |

```python
from collections import Counter
from sae_lens.loading.pretrained_saes_directory import get_pretrained_saes_directory

rows = [np_id for _, lookup in get_pretrained_saes_directory().items()
        for np_id in lookup.neuronpedia_id.values()
        if np_id and np_id.startswith("gemma-3-1b-it/")]
print(Counter(n.split("/", 1)[1].split("-", 1)[1] for n in rows))
```

Expected: **26** rows for `gemma-3-1b-it/gemmascope-2-transcoder-16k` and **18** for `-262k`. The
diff agrees exactly — 78 `neuronpedia:` lines added, 0 removed (26 for 1b-it-16k, 18 for 1b-it-262k,
34 for 4b-it-16k).

Reproducing the benchmark, and running the example notebook that consumes locally generated
dashboards, are documented once for the whole wave here:
**[Verifying dashboard generation](https://interpretune.readthedocs.io/en/latest/usage/verifying_dashboard_generation.html)**.
Cross-PR review order and the shared caveats live on the umbrella issue,
[interpretune#231](https://github.com/speediedan/interpretune/issues/231).

### CODE REVIEW

**Comment by chanind** [comment]
Hi @speediedan, thanks for the PR! Heads up: I've pushed a commit to this branch (11d0a0b) that reworks the `pretrained_saes.yaml` changes. Two issues with the previous version:

**1. Most of the added aliases don't exist on neuronpedia.org.** I checked every added alias against production Neuronpedia (each source page embeds `saelensRelease` / `saelensSaeId` / `hfFolderId`, and missing sources 404 on the feature API):

- `gemma-3-1b-it` has **no gemmascope-2 transcoder sources at all** — only `gemmascope-2-res-{16k,65k,262k}` at layers 7, 13, 17, 22. All 52 1b-it aliases pointed at pages that don't exist.
- `gemma-3-4b-it/N-gemmascope-2-transcoder-16k` only exists for layers 0–4, 6, 7, 9, 10, 12 — the other 24 layers 404.

I suspect these were validated against a local/hosted Neuronpedia instance with self-imported sources. The `neuronpedia:` field in SAELens needs to reference production neuronpedia.org, since it's used to link users to hosted dashboards.

**2. The branch was deleting 10 `- id:` lines rather than repairing them.** `main` has no pre-existing structural damage — every entry in `gemma-scope-2-1b-pt-transcoders-all` and `gemma-scope-2-12b-pt-transcoders-all` is intact. The deletions on this branch left orphaned `path:`/`l0:` lines that became duplicate YAML keys of the preceding entry, so e.g. `layer_15_width_16k_l0_small_affine` silently loaded `transcoder_all/layer_15_width_262k_l0_big` instead of its own weights, and the `layer_15_width_262k_l0_big` id disappeared entirely. Possibly a bad rebase — worth double-checking your fork's state against upstream main.

The yaml diff is now purely additive (+44): the 10 verified 4b-it 16k aliases, plus 34 `gemma-3-4b-it/N-gemmascope-2-transcoder-262k` aliases (all layers exist on production and map to `layer_N_width_262k_l0_small_affine`), which the original diff was missing.

The fold and activations-store commits are untouched — we'll review those separately.

**Comment by speediedan** [comment]
Thanks for catching the local Neuronpedia aliases and missing ids and merging @chanind! 

Checked, yes it was a bad rebase on the latter. For the former I'd originally planned a companion Neuronpedia PR to publish the `gemma-3-1b-it` transcoder sources, then pivoted to making them available as a pilot format for streaming dashboards via Hugging Face datasets and buckets instead and failed to revert the alias additions.

One minor unrelated thing that surfaced when I had Claude trace the provenance of the aforementioned bad links: 
There seems to be a full top-level
```yaml
  model: gpt2-small
  repo_id: neuronpedia/gpt2-small__res_scefr-ajt
  saes:
...
```
block nested underneath `id: blocks.10.hook_resid_pre` here (https://github.com/decoderesearch/SAELens/blob/695da861acb2b7e1d68bb1880f058d5b3aa013d6/sae_lens/pretrained_saes.yaml#L36214-L36232)

```yaml
gpt2-small-res_scefr-ajt:
  conversion_func: null
  links:
    dashboards: https://www.neuronpedia.org/gpt2-small/res_scefr-ajt
    model: https://huggingface.co/gpt2
  model: gpt2-small
  repo_id: neuronpedia/gpt2-small__res_scefr-ajt
  saes:
  - id: blocks.2.hook_resid_pre
    neuronpedia: gpt2-small/2-res_scefr-ajt
    path: 2-res_scefr-ajt
  - id: blocks.6.hook_resid_pre
    neuronpedia: gpt2-small/6-res_scefr-ajt
    path: 6-res_scefr-ajt
  - conversion_func: null
    id: blocks.10.hook_resid_pre
    links:
      dashboards: https://www.neuronpedia.org/gpt2-small/res_scefr-ajt
      model: https://huggingface.co/gpt2
    model: gpt2-small
    neuronpedia: gpt2-small/10-res_scefr-ajt
    path: 10-res_scefr-ajt
    repo_id: neuronpedia/gpt2-small__res_scefr-ajt
    saes:
    - id: blocks.2.hook_resid_pre
      neuronpedia: gpt2-small/2-res_scefr-ajt
      path: 2-res_scefr-ajt
    - id: blocks.6.hook_resid_pre
      neuronpedia: gpt2-small/6-res_scefr-ajt
      path: 6-res_scefr-ajt
    - id: blocks.10.hook_resid_pre
      neuronpedia: gpt2-small/10-res_scefr-ajt
      path: 10-res_scefr-ajt
```
I don't think it actually materially affects loading which is why it probably hasn't surfaced. Not worth a separate PR IMHO but might be cleanup worth rolling into your next related PR.

Thanks again for your continued contributions to SAELens and the open-source interpretability ecosystem! :rocket:
